### ISSUE TITLE
Session management is not enabled although session_state and check_session_iframe are set

### ISSUE THREAD

**Issue by hoalex** [pingidentity/mod_auth_openidc#276]
###### Environment

- mod_auth_openidc 2.2.0-1~jessie+1
- Apache 2.4.10-10+deb8u8
- Debian 8.8
- Keycloak 3.1.0.Final as the OpenID Connect Provider

###### Expected behaviour
I expect session management to be enabled when

- the OpenID Connect Provider sends a `"session_state"` value in the authentication response and
- the OIDC metadata includes a link to the `"check_session_iframe"`

###### Actual behaviour
Session management is not enabled by `mod_auth_openidc`, although Keycloak includes a `"session_state"` value in its authentication response and also includes a link to `"check_session_iframe"` in its metadata file.

###### Minimized example
Apache configuration:
```
<VirtualHost $IP:443>
	ServerName oidc.redacted.org
	SSLEngine On
	$SSL_CONFIG

	DocumentRoot /var/www/oidc/

	OIDCProviderMetadataURL https://idp.redacted.org/auth/realms/myrealm/.well-known/openid-configuration
	OIDCCryptoPassphrase 123
	OIDCRedirectURI https://oidc.redacted.org/cb
	OIDCClientId oidc
	OIDCClientSecret redacted
	LogLevel debug

	<Location />
	      AuthType openid-connect
	      Require valid-user
	</Location>

</VirtualHost>
```

Keycloaks metadata file (shortened):
```
{
    "issuer": "https://idp.redacted.org/auth/realms/myrealm",
    "authorization_endpoint": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/auth",
    "token_endpoint": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/token",
    "token_introspection_endpoint": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/token/introspect",
    "userinfo_endpoint": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/userinfo",
    "end_session_endpoint": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/logout",
    "jwks_uri": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/certs",
    "check_session_iframe": "https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/login-status-iframe.html",
}
```

###### Configuration and Log files
Debug log of the module during user authentication:
```
[...]
[auth_openidc:debug] [pid 11835] src/util.c(534): [client redacted:18138] oidc_util_http_call: url=https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/token, data=grant_type=authorization_code&code=BgLyvR1cmUYOyd85zq8TpouYDqWnd93R5W31j6Aa2nI.9e116a10-fd2b-42d7-94e9-4133434b919d&redirect_uri=https%3A%2F%2Foidc.redacted.org%2Fcb&state=h-nbyFIJu0XBBW6T2TQVl2LXyg4, content_type=application/x-www-form-urlencoded, basic_auth=oidc:redacted, bearer_token=(null), ssl_validate_server=1, referer: https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/auth?response_type=code&scope=openid&client_id=oidc&state=h-nbyFIJu0XBBW6T2TQVl2LXyg4&redirect_uri=https%3A%2F%2Foidc.redacted.org%2Fcb&nonce=sbVYfVUvLWMcbKyV1HhgVuA9jfTSM0MUru_nMOepQ7c
[auth_openidc:debug] [pid 11835] src/util.c(670): [client redacted:18138] oidc_util_http_call: HTTP response code=200, referer: https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/auth?response_type=code&scope=openid&client_id=oidc&state=h-nbyFIJu0XBBW6T2TQVl2LXyg4&redirect_uri=https%3A%2F%2Foidc.redacted.org%2Fcb&nonce=sbVYfVUvLWMcbKyV1HhgVuA9jfTSM0MUru_nMOepQ7c
[auth_openidc:debug] [pid 11835] src/util.c(675): [client redacted:18138] oidc_util_http_call: response={"access_token":"redacted","expires_in":300,"refresh_expires_in":1800,"refresh_token":"redacted","token_type":"bearer","id_token":"redacted","not-before-policy":0,"session_state":"7477281a-3c68-4a86-9b8d-2bc4c523bfc5"}, referer: https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/auth?response_type=code&scope=openid&client_id=oidc&state=h-nbyFIJu0XBBW6T2TQVl2LXyg4&redirect_uri=https%3A%2F%2Foidc.redacted.org%2Fcb&nonce=sbVYfVUvLWMcbKyV1HhgVuA9jfTSM0MUru_nMOepQ7c
[...]
[auth_openidc:warn] [pid 11835] [client redacted:18138] oidc_save_in_session: session management disabled: no "session_state" value is provided in the authentication response even though "check_session_iframe" (https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/login-status-iframe.html) is set in the provider configuration, referer: https://idp.redacted.org/auth/realms/myrealm/protocol/openid-connect/auth?response_type=code&scope=openid&client_id=oidc&state=h-nbyFIJu0XBBW6T2TQVl2LXyg4&redirect_uri=https%3A%2F%2Foidc.redacted.org%2Fcb&nonce=sbVYfVUvLWMcbKyV1HhgVuA9jfTSM0MUru_nMOepQ7c
```

**Comment by zandbelt** [comment]
The Session Management spec says:

http://openid.net/specs/openid-connect-session-1_0.html#CreatingUpdatingSessions

> When the OP supports session management, it MUST also return the Session State as an additional session_state parameter in the Authentication Response. The OpenID Connect Authentication Response is specified in Section 3.1.2.5 of OpenID Connect Core 1.0.

Section 3.1.2.5 of OpenID Connect Core 1.0 says:

http://openid.net/specs/openid-connect-core-1_0.html#AuthResponse

> An Authentication Response is an OAuth 2.0 Authorization Response message returned from the OP's Authorization Endpoint in response to the Authorization Request message sent by the RP.

> When using the Authorization Code Flow, the Authorization Response MUST return the parameters defined in Section 4.1.2 of OAuth 2.0 [RFC6749] by adding them as query parameters to the redirect_uri specified in the Authorization Request using the application/x-www-form-urlencoded format, unless a different Response Mode was specified.

Your log shows that a `session_state` value is provided as part of the response to a call to the token endpoint. That's the token response. The authentication response coming back from the authorization endpoint that happens before that however apparently did not contain it.

You'll need to make sure your OP returns `session_state` in the authorization response. I know Keycloak is capable of doing that.

**Comment by hoalex** [comment]
You're right, I looked at the wrong response!
Unfortunately, it appears that Keycloak only sends the "session_state" parameter in the authentication response if the response type includes "token" (see https://issues.jboss.org/browse/KEYCLOAK-4478).

Because I did not explicitly specify such a response type in the module, it defaulted to ``code`` which leads to the above mentioned behaviour. When chosing ``code token``, session management gets successfully enabled. I will try to get in contact with the Keycloak guys to get this solved.

As this is indeed not an error in mod_auth_openidc, I will close this ticket. Thanks for your help!
